### PR TITLE
Add basic usage example

### PR DESCRIPTION
**Pull request by ra0x3**
### Description

Whilst working on https://github.com/FuelLabs/fuel-indexer/issues/110, I noticed that we need a few more things in order to get a decently working block explorer backend, so I decided to just throw all the remaining pieces into this PR.

- This PR does  a few things:
  - Some general cleanup and organization.
  - Adds a new Basic Usage example -- which is a small block explorer backend
    - Thus addressing https://github.com/FuelLabs/fuel-indexer/issues/110
  - Adds the `Transaction` abstraction to `BlockData`
    - Previously we were using `Vec<Receipt>` to represent a `Transaction`, but after further looking into it, we actually do have the ability to get the actual `Transaction` via the Fuel GQL client
      - Now, `BlockData` is packed with `Transaction`s, and each `Transaction` contains its receipts `Vec<Receipt>`
  - Makes ABI JSON optional in the index manifest 
    - Thus addressing https://github.com/FuelLabs/fuel-indexer/issues/217
 - Allows foreign keys to be of any type 
   - Thus addressing https://github.com/FuelLabs/fuel-indexer/issues/225
 - Updates relevant docs
### Testing steps
- [ ] Unit tests were added for https://github.com/FuelLabs/fuel-indexer/issues/225 so CI should pass
- [ ] Spin up the block explorer example

```bash
# Clear database
bash scripts/utils/refresh_test_db.bash

# Start web components
bash scripts/utils/start_test_components.bash

# Start the indexer
cargo run --bin fuel-indexer -- --manifest examples/block-explorer/manifest.yaml

# Trigger some block/transactions
bash go.bash
```

### CODE REVIEW

**Review by deekerno — DISMISSED** [review]
Left some comments, but nothing blocking really.

File: examples/block-explorer/explorer-index/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+                        tokens_transferred.push(asset_id.to_string());
+                    }
+                    #[allow(unused)]
+                    Receipt::TransferOut {
+                        id,
+                        to,
+                        amount,
+                        asset_id,
+                        ..
+                    } => {
```
[the remark this answers is not in the document]
Comment (ra0x3) [reply examples/block-explorer/explorer-index/src/lib.rs:132]: @deekerno 
- Eventually. The PR is about 97% done -- I was running into that WASM issue I mentioned

```text
Oct 27 16:51:50.174 ERROR fuel_indexer::executor: WasmIndexExecutor handle_events failed: RuntimeError { source: Trap(UnreachableCodeReached), wasm_trace: [FrameInfo { module_name: "<module>", func_index: 286, function_name: Some("rust_panic"), func_start: SourceLoc(202485), instr: SourceLoc(202486) }, FrameInfo { module_name: "<module>", func_index: 189, function_name: Some("_ZN3std9panicking20rust_panic_with_hook17hb09154fa23e06c37E"), func_start: SourceLoc(199086), instr: SourceLoc(199167) }, FrameInfo { module_name: "<module>", func_index: 213, function_name: Some("_ZN3std9panicking19begin_panic_handler28_$u7b$$u7b$closure$u7d$$u7d$17h6091c197f0d08bf0E"), func_start: SourceLoc(200931), instr: SourceLoc(200971) }, FrameInfo { module_name: "<module>", func_index: 274, function_name: Some("_ZN3std10sys_common9backtrace26__rust_end_short_backtrace17h004afb3e6a867c40E"), func_start: SourceLoc(202363), instr: SourceLoc(202368) }, FrameInfo { module_name: "<module>", func_index: 251, function_name: Some("rust_begin_unwind"), func_start: SourceLoc(202064), instr: SourceLoc(202075) }, FrameInfo { module_name: "<module>", func_index: 212, function_name: Some("_ZN4core9panicking9panic_fmt17h9e229748e3ae9f9dE"), func_start: SourceLoc(200867), instr: SourceLoc(200925) }, FrameInfo { module_name: "<module>", func_index: 14, function_name: Some("handle_events"), func_start: SourceLoc(91500), instr: SourceLoc(94436) }], native_trace:    0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
   4: <unknown>
   5: <unknown>
   6: <unknown>
   7: <unknown>
   8: <unknown>
   9: <unknown>
  10: <unknown>
  11: <unknown>
  12: <unknown>
  13: <unknown>
  14: <unknown>
  15: <unknown>
 }.
```

- But for the life of me can't figure out what the issue is - especially when the `fuel_indexer_test.wasm` module seems to work fine (as evidenced by CI passing)
- I'm gonna just wait until the WASM issue is figured out until merging this - but still wanted to get your eyes on it in the mean time 🆗

[the remark this answers is not in the document]
Comment (deekerno) [reply examples/block-explorer/explorer-index/src/lib.rs:132]: I haven't been able to find out the source of that error either, but it goes away when I re-build the WASM module. 🤷🏽 I've been doing a bit of a deep-dive on WASM runtimes and how to debug errors more efficiently, so hopefully there's some info out there on how to make it easier to work with.

[the remark this answers is not in the document]
Comment (ra0x3) [reply examples/block-explorer/explorer-index/src/lib.rs:132]: - For any who come by this PR:
  - The issue is resolved - and was related to build caching
  - WASM is still a black box 🗳️

**Review by deekerno — APPROVED** [review]

**Pull request merged**